README.md
# vocoder for clojure's Overtone

   It makes robot-y vocals, like this: 
  
  https://www.youtube.com/watch?v=BSsXYisIE4E

   Vocoder emulator, designed to be attached as an fx instance to an
   instrument or to other synths via bus.

   Uses amplitude of frequency bands in an input signal (usually the mic)
   to control the amplitude of frequency bands in a synthesiser.

   Number of bands, frequency range and number of passes through the bpf
   can be adjusted when the vocoder synth is defined (but not 'live' using 
   'ctl').

   This emulator also includes a white-noise synthesizer for picking up 
   higher frequencies in speech (called sibilants).

   The thing with vocoders is that the synth settings are just as important
   as the vocoder's. You will need to experiment with your synth to get the
   the sounds you want, but these defaults work well to give a 'roboty'
   sound using the pad synth in overtone.inst.synth library (you might need
   to tweak it to gate properly though, see example folder).

   Enjoy and send feedback to @nseymoursmith on twitter and github.

## License

Copyright © 2016 nseymoursmith

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-01-28
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-01-28
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/vocoder/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/vocoder/compare/0.1.0...0.1.1


test/vocoder/core_test.clj
(ns vocoder.core-test
  (:require [clojure.test :refer :all]
            [vocoder.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


example/vocoder_example.clj
(ns vocoder.vocoder-example
  (:use [overtone.live]
        [vocoder.core]))
(vocoder)
;;This synth works well with the vocoder presets
(definst pad2
  [note 60 t 10 amt 0.3 amp 0.1 a 0.4 d 0.5 s 0.8 r 2 gate 1]
  (let [freq   (midicps note)
        lfo    (+ 2 (* 0.01 (sin-osc:kr 5 (rand 1.5))))
        src    (apply + (saw [freq (* freq lfo)]))
        env    (env-gen (adsr a d s r) gate :action FREE)
        f-env  (x-line:kr 0.001 4 t)
        src    (* env src)
        signal (rlpf src (+ (* 0.3 freq) (* f-env 2 freq)) 0.5)
        k      (/ (* 4 amt) (- 1 amt))
        dist   (clip2 (/ (* (+ 1 k) signal) (+ 1 (* k (abs signal))))
                      0.03)
        snd    (* amp dist (line:kr 1 0 t))]
    src))

;;Attach vocoder to synth and start midi player
(do
  (def voco (inst-fx! pad2 vocoder))
  (def voc-player (midi-poly-player pad2))
)

;;Or attach vocoder to synth and just play a note
;;and play using ctl
(do
  (def voco (inst-fx! pad2 vocoder))
  (pad2)
)
(ctl pad2 :note 50)

;;Stop midi, kill fx and synths
(do
  (clear-fx pad2)
  (midi-player-stop voc-player)
  (stop)
)

;;recording
(recording-start "~/Desktop/mic-check-24.wav")
(recording-stop)

;;preset play
(do 
  (ctl voco :comp-thresh 0.01)
  (ctl voco :voc-gain 1)
  (ctl voco :src-gain 1)
  (ctl voco :comp-gain 1.0)
  (ctl voco :fsib 8000)
  (ctl voco :sib-gain 0.1)
  (ctl voco :rq-voc 0.32)
  (ctl voco :rq-synth 0.32)
  (ctl voco :distance 10)
  (ctl voco :gain 140.0))


src/vocoder/core.clj
(ns vocoder.core
  (:use [overtone.live]))

(defn- nbpf
  "n-stage recursive band-pass filter"
  [input freq rq stages]
  (loop [n stages signal input]
    (if (zero? n) signal
        (recur (dec n) (bpf signal freq rq)))))

(defn- log-spacing
  "List of numbers, evenly spaced on a logarithmic scale, 
  'end' value inclusive"
  [start end steps]
  (let [base (Math/pow (/ end start) (/ steps))
        n (range steps)]
    (map #(* start (Math/pow base (+ % 1))) n)))

(defsynth vocoder
  "
   Vocoder emulator, designed to be attached as an fx instance to an
   instrument or to other synths via bus.

   Uses amplitude of frequency bands in an input signal (usually the mic)
   to control the amplitude of frequency bands in a synthesiser.

   Number of bands, frequency range and number of passes through the bpf
   can be adjusted when the vocoder synth is defined (but not 'live' using 
   'ctl').

   This emulator also includes a white-noise synthesizer for picking up 
   higher frequencies in speech (called sibilants).

   The thing with vocoders is that the synth settings are just as important
   as the vocoder's. You will need to experiment with your synth to get the
   the sounds you want, but these defaults work well to give a 'roboty'
   sound using the pad synth in overtone.inst.synth library (you might need
   to tweak it to gate rather than have a perc envelope though).

   Enjoy and send feedback to @nseymoursmith on twitter and github.
   "
  [bus 0 mic-input 8
   comp-thresh 0.01 comp-slope 5
   fmin 20.0 fmax 10000.0 nbands 9 stages 2 rq-voc 0.32 rq-synth 0.32
   distance 10.0
   voc-gain 1.0 src-gain 1.0 combination-gain 1.0
   fsib 8000.0 sib-gain 0.1
   gain 140.0]
  (let [; - sources - ;
        src (in bus)
        input (in mic-input)

        ; - mic compression and analysis -
        compressed (compander :in input 
                              :control input 
                              :thresh comp-thresh 
                              :slope-below comp-slope)
        voc (* voc-gain compressed)
        voc-level (amplitude:kr voc 0.1 0.1)
        band-freqs (log-spacing (. fmin value) (. fmax value) (. nbands value))
        voc-bands (map #(nbpf voc % rq-voc (. stages value)) band-freqs)
        taus (map #(/ distance %) band-freqs)
        voc-amps (map (fn [band tau] (amplitude band tau tau)) 
                      voc-bands taus)

        ; - extract and mix src bands -
        src (* src-gain src)
        src-level (amplitude:kr src 0.1 0.1)
        src-bands (map #(nbpf src % rq-synth (. stages value)) band-freqs)
        src-voc-mix (map * src-bands voc-amps)
        composited (* combination-gain (sum src-voc-mix))

        ; - white-noise synthesis for sibilance - ;
        a (buffer 2048)
        b (buffer 2048)
        noise-source (white-noise)
        formed (pv-mul (fft a compressed) (fft b noise-source))
        sib-synth (ifft formed)
        sibilance (hpf sib-synth fsib)
        sibilance (* sib-gain src-level sibilance)

        output (sum [composited sibilance])]
    (replace-out bus (* gain output))))

